File: clients/ruby/model/projectile.rb
require './model/vec2'

module Model

# Weapon projectile
class Projectile
    # Unique id
    attr_accessor :id
    # Index of the weapon this projectile was shot from (starts with 0)
    attr_accessor :weapon_type_index
    # Id of unit who made the shot
    attr_accessor :shooter_id
    # Id of player (team), whose unit made the shot
    attr_accessor :shooter_player_id
    # Current position
    attr_accessor :position
    # Projectile's velocity
    attr_accessor :velocity
    # Left time of projectile's life
    attr_accessor :life_time

    def initialize(id, weapon_type_index, shooter_id, shooter_player_id, position, velocity, life_time)
        @id = id
        @weapon_type_index = weapon_type_index
        @shooter_id = shooter_id
        @shooter_player_id = shooter_player_id
        @position = position
        @velocity = velocity
        @life_time = life_time
    end

    # Read Projectile from input stream
    def self.read_from(stream)
        id = stream.read_int()
        weapon_type_index = stream.read_int()
        shooter_id = stream.read_int()
        shooter_player_id = stream.read_int()
        position = Model::Vec2.read_from(stream)
        velocity = Model::Vec2.read_from(stream)
        life_time = stream.read_double()
        Projectile.new(id, weapon_type_index, shooter_id, shooter_player_id, position, velocity, life_time)
    end

    # Write Projectile to output stream
    def write_to(stream)
        stream.write_int(@id)
        stream.write_int(@weapon_type_index)
        stream.write_int(@shooter_id)
        stream.write_int(@shooter_player_id)
        @position.write_to(stream)
        @velocity.write_to(stream)
        stream.write_double(@life_time)
    end

    def to_s
        string_result = "Projectile { "
        string_result += "id: "
        string_result += @id.to_s
        string_result += ", "
        string_result += "weapon_type_index: "
        string_result += @weapon_type_index.to_s
        string_result += ", "
        string_result += "shooter_id: "
        string_result += @shooter_id.to_s
        string_result += ", "
        string_result += "shooter_player_id: "
        string_result += @shooter_player_id.to_s
        string_result += ", "
        string_result += "position: "
        string_result += @position.to_s
        string_result += ", "
        string_result += "velocity: "
        string_result += @velocity.to_s
        string_result += ", "
        string_result += "life_time: "
        string_result += @life_time.to_s
        string_result += " }"
        string_result
    end

    def to_str
        to_s
    end
end

end

File: clients/typescript/src/model/projectile.ts
import { Vec2 } from "./vec2";
import { Stream } from "../stream";

/**
 * Weapon projectile
 */
export class Projectile {
    /**
     * Unique id
     */
    id: number
    /**
     * Index of the weapon this projectile was shot from (starts with 0)
     */
    weaponTypeIndex: number
    /**
     * Id of unit who made the shot
     */
    shooterId: number
    /**
     * Id of player (team), whose unit made the shot
     */
    shooterPlayerId: number
    /**
     * Current position
     */
    position: Vec2
    /**
     * Projectile's velocity
     */
    velocity: Vec2
    /**
     * Left time of projectile's life
     */
    lifeTime: number

    constructor(id: number, weaponTypeIndex: number, shooterId: number, shooterPlayerId: number, position: Vec2, velocity: Vec2, lifeTime: number) {
        this.id = id;
        this.weaponTypeIndex = weaponTypeIndex;
        this.shooterId = shooterId;
        this.shooterPlayerId = shooterPlayerId;
        this.position = position;
        this.velocity = velocity;
        this.lifeTime = lifeTime;
    }

    /**
     * Read Projectile from input stream
     */
    static async readFrom(stream: Stream): Promise<Projectile> {
        let id;
        id = await stream.readInt();
        let weaponTypeIndex;
        weaponTypeIndex = await stream.readInt();
        let shooterId;
        shooterId = await stream.readInt();
        let shooterPlayerId;
        shooterPlayerId = await stream.readInt();
        let position;
        position = await Vec2.readFrom(stream);
        let velocity;
        velocity = await Vec2.readFrom(stream);
        let lifeTime;
        lifeTime = await stream.readDouble();
        return new Projectile(id, weaponTypeIndex, shooterId, shooterPlayerId, position, velocity, lifeTime)
    }

    /**
     * Write Projectile to output stream
     */
    async writeTo(stream: Stream) {
        let id = this.id;
        await stream.writeInt(id);
        let weaponTypeIndex = this.weaponTypeIndex;
        await stream.writeInt(weaponTypeIndex);
        let shooterId = this.shooterId;
        await stream.writeInt(shooterId);
        let shooterPlayerId = this.shooterPlayerId;
        await stream.writeInt(shooterPlayerId);
        let position = this.position;
        await position.writeTo(stream);
        let velocity = this.velocity;
        await velocity.writeTo(stream);
        let lifeTime = this.lifeTime;
        await stream.writeDouble(lifeTime);
    }
}